product_contract/models/sale_order_line.py

```diff
@@ -0,0 +1,279 @@
+            rec.date_end = rec._get_date_end() if rec.date_start else False
+
+    def _get_contract_line_qty(self):
+        """Returns the quantity to be put on new contract lines."""
+        self.ensure_one()
+        # The quantity on the generated contract line is 1, as it
+        # correspond to the most common use cases:
+        # - quantity on the SO line = number of periods sold and unit
+        #   price the price of one period, so the
+        #   total amount of the SO corresponds to the planned value
+        #   of the contract; in this case the quantity on the contract
+        #   line must be 1
+        # - quantity on the SO line = number of hours sold,
+        #   automatic invoicing of the actual hours through a variable
+        #   quantity formula, in which case the quantity on the contract
+        #   line is not used
+        # Other use cases are easy to implement by overriding this method.
+        return 1.0
+
+    def _prepare_contract_line_values(
+        self, contract, predecessor_contract_line_id=False
+    ):
+        """
+        :param contract: related contract
+        :param predecessor_contract_line_id: contract line to replace id
+        :return: new contract line dict
+        """
+        self.ensure_one()
+        recurring_next_date = self.env[
+            "contract.line"
+        ]._compute_first_recurring_next_date(
+            self.date_start or fields.Date.today(),
+            self.recurring_invoicing_type,
+            self.recurring_rule_type,
+            1,
+        )
+        termination_notice_interval = self.product_id.termination_notice_interval
+        termination_notice_rule_type = self.product_id.termination_notice_rule_type
+        return {
+            "sequence": self.sequence,
+            "product_id": self.product_id.id,
+            "name": self.name,
+            "quantity": self._get_contract_line_qty(),
+            "uom_id": self.product_uom.id,
+            "price_unit": self.price_unit,
+            "discount": self.discount,
+            "date_end": self.date_end,
+            "date_start": self.date_start or fields.Date.today(),
+            "recurring_next_date": recurring_next_date,
+            "recurring_interval": 1,
+            "recurring_invoicing_type": self.recurring_invoicing_type,
+            "recurring_rule_type": self.recurring_rule_type,
+            "is_auto_renew": self.is_auto_renew,
+            "auto_renew_interval": self.auto_renew_interval,
+            "auto_renew_rule_type": self.auto_renew_rule_type,
+            "termination_notice_interval": termination_notice_interval,
+            "termination_notice_rule_type": termination_notice_rule_type,
+            "contract_id": contract.id,
+            "sale_order_line_id": self.id,
+            "predecessor_contract_line_id": predecessor_contract_line_id,
+            "analytic_account_id": self.order_id.analytic_account_id.id,
```

Comment:
There is no more `analytic_account_id`, now you need to use `analytic_distribution`.

Reply:
Hi @Rad0van thank you for porting this module, I just tested it locally (cannot find that PR on Runboat...) and except for this missing analytic distribution, it works fine !

analytic_account_id still exists on sale order, although analytic_distribution from sale order line needs to be forwarded to contract line and then to invoice line.
You should add the following :
```suggestion
            "analytic_account_id": self.order_id.analytic_account_id.id,
            "analytic_distribution": self.analytic_distribution,
```

Reply:
@remi-filament done.